src/main/java/ch/njol/skript/expressions/ExprItems.java

```diff
@@ -0,0 +1,146 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.expressions;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.aliases.ItemType;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.Literal;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.util.Utils;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.iterator.IteratorIterable;
+import com.google.common.collect.Iterables;
+import org.bukkit.Material;
+import org.bukkit.event.Event;
+import org.bukkit.inventory.ItemStack;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.StreamSupport;
+
+@Name("Items")
+@Description("Items or blocks of a specific type, useful for looping.")
+@Examples({
+	"loop items of type ore and log:",
+		"\tblock contains loop-item",
+		"\tmessage \"Theres at least one %loop-item% in this block\"",
+		"drop all blocks at the player # drops one of every block at the player"
+})
+@Since("1.0 pre-5")
+public class ExprItems extends SimpleExpression<ItemType> {
+
+	private static final ItemType[] ALL_BLOCKS = Arrays.stream(Material.values())
+		.filter(Material::isBlock)
+		.map(ItemType::new)
+		.toArray(ItemType[]::new);
+
+	static {
+		Skript.registerExpression(ExprItems.class, ItemType.class, ExpressionType.COMBINED,
+			"[all [[of] the]|the] block[[ ]type]s",
+			"every block[[ ]type]",
+			"[all [[of] the]|the|every] block[s] of type[s] %itemtypes%",
+			"[all [[of] the]|the|every] item[s] of type[s] %itemtypes%"
+		);
+	}
+
+	@Nullable
+	private Expression<ItemType> itemTypeExpr;
+	private boolean items;
+	private ItemType[] buffer = null;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		items = matchedPattern == 3;
+		itemTypeExpr = matchedPattern == 0 || matchedPattern == 1 ? null : (Expression<ItemType>) exprs[0];
+		if (itemTypeExpr instanceof Literal) {
+			for (ItemType itemType : ((Literal<ItemType>) itemTypeExpr).getAll())
+				itemType.setAll(true);
+		}
+		return true;
+	}
+
+	@Override
+	@Nullable
+	protected ItemType[] get(Event event) {
+		if (buffer != null)
+			return buffer;
+		List<ItemType> items = new ArrayList<>();
+		iterator(event).forEachRemaining(items::add);
+		if (itemTypeExpr instanceof Literal)
+			return buffer = items.toArray(new ItemType[0]);
```

Comment:
can we just assign the var here? the return below does the same thing and personally I think using assignment expressions like this can be easy to overlook